### DISCUSSION TITLE
Unable to run programs off the hub

### DISCUSSION THREAD

**Discussion by JJHackimoto** [pybricks/support#1373] [Q&A] [answered]
Hi,

I just started using Pybricks Block coding for a complicated bowling pinsetter build, that uses three technic hubs that communicate with each other.
I have transferred over my program by hand from the PoweredUp app, and divided it into three separate programs, one for each hub. The hubs then send values to each other using the broadcasting feature.
I've installed the firmware on all three hubs, downloaded and ran the program (running the program doesn't do anything at this point since they all just wait for commands from each other), and I've then turned the hubs off.

Now the problem.  When I turn the hubs back on, they all just search for a bluetooth connection (rapid doubleblinking blue light). Pressing the green button once during this process does nothing. I expected the program to run and the light to fade in and out, but that's not the case. My expectation is based on this: https://pybricks.com/install/technic-boost-city/

Connecting them all from separate chrome tabs doesn't work either. When trying to run a program with more than one hub connected, it hangs when downloading to the hub at 48%. I also have the following error in the Chrome inspector, but idk if that has anything to do with this: "DOMException: GATT operation failed for unknown reason."

I'm sure the Pybricks firmware is installed since the blue flashing doesn't stop after the same amount of time as the original firmware did when not touching it.

Does anyone have a clue as to what could be going on?

Hardware:
Three Lego Technic hubs running the latest stable beta version (downloaded from the beta since I use block coding)
M1 Pro MacBook Pro running Google Chrome

**Comment by laurensvalk** [comment]
To make sure it's not an issue with your program, can you try running something like the following?

This should help you verify whether starting it with the button works or not.

![image](https://github.com/pybricks/support/assets/12326241/74e7f126-1a10-41b5-9efe-a1cafcde190a)

It's possible that the program stopped for another reason. For example, maybe one of your motors wasn't plugged in?

It's often most practical to do this sort of test while you're still connected to your computer. That way, you can see that kind of errors when they occur.

-------------

When using Hub to Hub communication, it's important to note that the `observe` command will report `None` if there is no data from another hub yet. If you'd use `None` to set the speed of a motor, for example, you'd get an error and the program would stop. We may be able to give more suggestions if you share your program :)

**Reply by JJHackimoto** [reply]
Thanks for your quick reply!

It's a bit embarrassing that I never tried such a simple program before posting here. Running the program you suggested does indeed work, which means that there's a problem in my specific code.

I'm not sure if I'm supposed to ask for solutions here but I'll try.

Coming from Lego PoweredUp, Ev3 and NXT programming, I expected the console to not be used when in block coding mode, resulting in me missing the errors I got there. I didn't think block coding could produce errors and instead assumed it wasn't possible to do anything "the wrong way".  Also, the errors are quite difficult to understand since they just point to lines in the Python code, that I tried my best at understanding without success.

Anyways, here's the beginning of my program (both in block-form and in Python form to trace the error. As seen in the screenshot, I don't even need to have any program section for this to occur. I also tried not to declare the color white, and instead used the built in "blue" for the DistributorSensor block, but that threw the same error.

<img width="492" alt="Skärmavbild 2024-01-11 kl  19 02 14" src="https://github.com/pybricks/support/assets/31407689/85d33c33-45db-4e22-958d-d18e0cd4e7de">
<img width="559" alt="Skärmavbild 2024-01-11 kl  19 02 44" src="https://github.com/pybricks/support/assets/31407689/e281d159-18a0-49ba-a8b0-35de7c73732f">
<img width="393" alt="Skärmavbild 2024-01-11 kl  19 03 07" src="https://github.com/pybricks/support/assets/31407689/39f957bb-c041-467a-98a9-f5f90bd7122f">

**Reply by laurensvalk** [reply]
Thanks for posting this. This is an error that's on our side, though something you can fix in your program too.

You've set up the color sensor to detect only white and nothing else. This is probably not what you intended :)

But the code isn't expecting just a single color, and throws that error. That's something we can fix on our end, so at least it won't fail.

> they just point to lines in the Python code

You're right, this is one the things on our list to improve.

**Reply by JJHackimoto** [reply]
Thank you! That fixed the error.

My suggestion to this would be that the Color Sensor setup block either has a minimum of two colors to be set (but this can still be difficult to understand if only one color is what you want to use), or to have for example "none" be a hidden default when only selecting a single color. This would then be visible only in the python code, but not in block coding. I'm sure you'll find a good solution. :)

In my example, I'm telling the sensor to look for white bowling pins. I basically just want it to look for white and nothing else would be interesting in this case. I have now set the second value to "none" since there's plenty of empty space in front of the sensor until a bowling pin appears. I'll see if that works better. :)

However, the hub now crashes when I try to run the program. (It turns off right away) No errors appear in the console either.
I'll just post the entire .py file here, but in .txt format since .py doesn't seem to be allowed here (I'm not able to select it in the file upload). I have no clue where to start troubleshooting this without errors.
[DriveHub.txt](https://github.com/pybricks/support/files/13907405/DriveHub.txt)

**Reply by laurensvalk** [reply]
> In my example, I'm telling the sensor to look for white bowling pins. I basically just want it to look for white and nothing else would be interesting in this case.

This is the part we have to clarify somehow. This question comes up a lot :)

The only way to detect something, is to distinguish it from something that it is not. If this sounds philosophical, look at it this way. How can you tell what a bowling pin looks like, if all you could possibly ever see is a bowling pin? You'd also need to know what *isn't* a bowling pin :smile: So indeed, you'll want to define another color called "background", which it whatever it sees when there isn't a pin.

We have a video on trains about it [here](https://www.youtube.com/watch?v=kwnSbnlhOkA). Notice that we need to tell it not just red/blue as the colors we want to see, but also what the track/backround looks like.

> My suggestion to this would be that the Color Sensor setup block either has a minimum of two colors to be set

Great idea!

**Reply by BertLindeman** [reply]
> However, the hub now crashes when I try to run the program. (It turns off right away) No errors appear in the console either.

I can confirm this. 
Your program starts and then the light on the Technic hub dims and the light goes out.
Running a technichub version `('technichub', '3.3.0', 'v1.20.0-23-g6c633a8dd on 2023-11-24')`

Some things I did after the original test in the python code:
 - removed all ```await wait(0)``` as I do not understand the use of it.
 - changed all ```await wait(1)``` into ```await wait(1)``` just to give the program a bit more time to do what is needed
 - added a print 'function name' to the start of all functions. 

This all does not change anything. Except a few prints.

Next I removed almost all code from the definition of ```main2``` and ```main3```.
That results in an error message. 
Here is a screenshot of it:
![afbeelding](https://github.com/pybricks/support/assets/8142081/afc5f38e-ae35-4488-b164-f89988655812)

Hope this helps.

Bert

**Reply by laurensvalk** [reply]
Thanks Bert! I must have missed the part about the crash. Would you mind posting just the Python code to a dedicated issue?

Can it be reproduced with just the devices attached as described?

**Reply by laurensvalk** [reply]
Maybe it's the same issue as https://github.com/pybricks/support/issues/1357.

The Technic Hub doesn't seem to like multitasking combined with Bluetooth.

**Reply by JJHackimoto** [reply]
Thanks Bert for checking this! I'll try to rearrange the program to avoid it using broadcasting while multitasking, if possible, while waiting for the bug to be fixed. :)

**Reply by laurensvalk** [reply, marked as answer]
We released a new beta version that should fix this. You can try it out with the new firmware at https://beta.pybricks.com/

If you've used that page before, it may take a few minutes for the latest version to load in the background. If you go to settings/about, you'll see 3.4.0b1 as the new version. When you do, you can go ahead and update the firmware or try new features.

![image](https://github.com/pybricks/support/assets/12326241/f7a38ea8-df61-4737-a4e6-0424bd4b3876)

**Reply by JJHackimoto** [reply]
Thank you!

I had to do the installation process about five times for each hub since it kept timing out with the following error: "The hub took too long to respond. Restart the hub and try again.". After that, they all started fine and no crashes occurred!

I marked your reply as the answer since it solved my initial problem. However, I've now done some testing and there's still a bug present here. Broadcasting and receiving values work fine, however, the variables won't update in the entire program at once. See my simple program attached. I have two parallell programs running at different intervals. One hub sends the "TriggerDistributor" value (true) by broadcasting. This value won't be changed for about three seconds or so, and then go back to being false. This is picked up by the first print function and shows in the console, however the other print function in the lower program will always print false. This is a problem since I'm using these values in different parts of the program, usually with "wait until" blocks, waiting until the variable becomes true.

If this is an actual bug, do you want me to create a separate issue for it?
<img width="724" alt="Skärmavbild 2024-02-01 kl  22 26 48" src="https://github.com/pybricks/support/assets/31407689/3d3aedd7-d62e-4ae0-9cf3-a1a7a24f661a">

**Reply by laurensvalk** [reply]
Good catch! Yes, I think I see where the bug might be. Feel free to create a dedicated issue for it. Thanks!

I think the global keyword isn’t included properly in the generated code.

**Comment by BertLindeman** [comment]
> However, the hub now crashes when I try to run the program. (It turns off right away) No errors appear in the console either.



> Maybe it's the same issue as #1357.
> 
> The Technic Hub doesn't seem to like multitasking combined with Bluetooth.

Tested the original program and commented-out the observe and broadcast statements.
No more crash seen then.
So looks like #1357 indeed.

[EDIT] Added a link to here in #1357 (Thought those were made automagically)

**Reply by laurensvalk** [reply]
It's probably due to:

https://github.com/pybricks/pybricks-micropython/blob/master/pybricks/common/pb_type_ble.c#L430-L443

So we need to fix https://github.com/pybricks/support/issues/1096.

Prediction came true :smile: 

> Due to a new upcoming editor feature, I think multi-tasking is going to get a lot more testing fairly soon. I kind of suspect there will still be some issues to resolve when lots of multitasking is being used, and this is probably one of them.

_Originally posted by @laurensvalk in https://github.com/pybricks/support/issues/1096#issuecomment-1821112329_

**Reply by dlech** [reply]
There is a pull request to fix this if anyone wants to try out the firmware there. https://github.com/pybricks/pybricks-micropython/pull/216
